types/project.go

```diff
@@ -185,13 +188,19 @@ func (p *Project) AllServices() Services {
 
 type ServiceFunc func(name string, service ServiceConfig) error
 
-// WithServices run ServiceFunc on each service and dependencies according to DependencyPolicy
-func (p *Project) WithServices(names []string, fn ServiceFunc, options ...DependencyOption) error {
+// WithServices runs ServiceFunc on each service and dependencies according to DependencyPolicy
+// It returns a new Project instance with the changes and keep the original Project unchanged
+func (p *Project) WithServices(names []string, fn ServiceFunc, options ...DependencyOption) (*Project, error) {
```

Comment:
This method is the only one I think is odd - the returned project option is always identical to the original.

I think it might make sense to have a pair of functions:
 * `ForEach(func(name string, svc ServiceConfig) error) error` - functional loop helper, does not copy project
 * `MutateServices(func(name string, svc *ServiceConfig) error) (*Project, error)` - clone project, allowing caller to get a mutable reference to each service in the process

Reply:
Agree. As we now use `WithXX` fr mutating functions, this name is not relevant anymore. `ForEach` would be a better one. We then can have a new `WithServiceFn(func(ServiceConfig) ServiceConfig)` for service mutation, if relevant (afaik we don't use this in docker/compose)